crates/oxc_linter/src/plugin.rs

```diff
@@ -0,0 +1,233 @@
+
+enum SpanInfo {
+    SingleSpanInfo(SingleSpanInfo),
+    MultipleSpanInfo(MultipleSpanInfo),
+}
+
+#[derive(Debug, Deserialize)]
+struct SingleSpanInfo {
+    span_start: u64,
+    span_end: u64,
+}
+
+#[derive(Debug, Deserialize)]
+struct MultipleSpanInfo {
+    span_start: Box<[u64]>,
+    span_end: Box<[u64]>,
+}
+
+#[derive(Deserialize, Clone)]
+pub struct InputQuery {
+    pub name: String,
+    pub query: String,
+    pub args: BTreeMap<Arc<str>, TransparentValue>,
+    pub reason: String,
+    #[serde(default)]
+    pub tests: QueryTests,
+}
+
+#[derive(Deserialize, Default, Clone)]
+pub struct QueryTests {
+    pub pass: Vec<SingleTest>,
+    pub fail: Vec<SingleTest>,
+}
+
+#[derive(Deserialize, Clone)]
+pub struct SingleTest {
+    pub relative_path: Vec<String>,
+    pub code: String,
+}
+
+pub struct LinterPlugin {
+    pub(super) rules: Vec<InputQuery>,
+    schema: &'static Schema,
+}
+
+impl LinterPlugin {
+    pub fn new(schema: &'static Schema) -> Self {
+        let queries_path = env::var("OXC_PLUGIN").unwrap();
+        let rules = fs::read_dir(queries_path)
+            .expect("to readdir queries_path folder")
+            .filter_map(std::result::Result::ok)
+            .filter(|dir_entry| dir_entry.path().is_dir())
+            .filter(|dir| !dir.path().to_str().map_or_else(|| true, |f| f.contains("ignore_")))
+            .filter_map(|folder| fs::read_dir(folder.path()).ok())
+            .flat_map(std::iter::IntoIterator::into_iter)
+            .filter_map(std::result::Result::ok)
+            .filter(|dir_entry| dir_entry.path().is_file())
+            .filter(|f| {
+                std::path::Path::new(f.path().as_os_str().to_str().unwrap())
+                    .extension()
+                    .map_or(false, |ext| ext.eq_ignore_ascii_case("yml"))
+            })
+            .map(|f| fs::read_to_string(f.path()))
+            .map(std::result::Result::unwrap)
+            .map(|rule| {
+                serde_yaml::from_str::<InputQuery>(rule.as_str())
+                    .expect(&format!("{rule}\n\nQuery above"))
+            })
+            .collect::<Vec<_>>();
+
+        Self { rules, schema }
+    }
+
+    pub fn run(
+        &self,
+        ctx: &mut LintContext,
+        semantic: Rc<Semantic<'_>>,
+        relative_file_path_parts: Vec<Option<String>>,
+    ) {
+        let inner = Adapter::new(semantic, relative_file_path_parts);
+        let adapter = Arc::from(&inner);
+        for input_query in &self.rules {
+            for data_item in execute_query(
+                self.schema,
+                Arc::clone(&adapter),
+                &input_query.query,
+                input_query.args.clone(),
+            )
+            .expect(
+                format!("not a legal query in query: \n\n\n{}", input_query.query.as_str())
+                    .as_str(),
+            )
+            .map(|v| {
+                if env::var("OXC_PRINT_TFALL_OUTPUTS").unwrap_or_else(|_| "false".to_owned())
```

Comment:
do we want to keep this? I think it's useful for testing, but when we make a playground this might be unnecessary

Reply:
would be nice to not abbreviate Trustfall to TFALL

Reply:
Fixed!